packages/agents/src/agents/Chameleon/index.ts

```diff
@@ -86,20 +100,87 @@ export class ChameleonAgent extends Agent {
         },
         this.config.prompts.loopPreventionPrompt,
         this.config.prompts.agentSpeakPrompt,
-        this.beforeLlmResponse.bind(this)
+        this.beforeLlmResponse2.bind(this)
       );
     } catch (err) {
       this.context.logger.error(err);
       return ResultErr("Unrecoverable error encountered.");
     }
   }
 
-  protected initializeChat(args: GoalRunArgs): void {
+  protected initializeChat(): void {
     const { chat } = this.context;
-    
+
     chat.persistent(buildDirectoryPreviewMsg(this.context.workspace));
     chat.persistent("user", prompts.exhaustAllApproaches);
-    chat.persistent("user", args.goal);
+  }
+
+  protected async beforeLlmResponse2(): Promise<{ logs: ChatLogs, agentFunctions: FunctionDefinition[], allFunctions: AgentFunction<AgentContext>[]}> {
+
+    // Retrieve the last message
+    const rawChatLogs = this.context.chat.chatLogs;
+    const lastMessages = [
+      rawChatLogs.getMsg("persistent", -1),
+      rawChatLogs.getMsg("temporary", -2),
```

Comment:
Because this is the "raw" chat log, it's append-only and is never manipulated. The "contextualized chat" (cChat) returns a new chat log based on this raw log. So I think we can safely assume these last messages are the most recent.